File: Easy/496_NextGreaterElementI.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
const nextGreaterElement = (nums1, nums2) => {
    let stack = [];
    let map = {};
    let result = [];
    stack.push(nums2[0]);

    for (let i = 1; i < nums2.length; i++) {
        while (stack.length > 0 && nums2[i] > stack[stack.length - 1]) {
            map[stack[stack.length - 1]] = nums2[i];
            stack.pop();
        }
        stack.push(nums2[i]);
    }

    for (const num of stack) {
        map[num] = -1;
    }

    for (const num of nums1) {
        result.push(map[num]);
    }

    return result;
};

File: Easy/496_next-greater-element-i.java
package Easy;

class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        // key: element itself; value: the first larger element in the right
        Map<Integer, Integer> map = new HashMap<>();

        Deque<Integer> stack = new ArrayDeque<>();
        for (int i = 0; i < nums2.length; i++) {
            while (!stack.isEmpty() && nums2[stack.peek()] < nums2[i]) {
                int index = stack.pop();
                map.put(nums2[index], nums2[i]);
            }

            stack.push(i);
        }

        // find the corresponding values for nums1[i] in map
        int[] result = new int[nums1.length];
        Arrays.fill(result, -1);
        for (int i = 0; i < nums1.length; i++) {
            int key = nums1[i];
            if (map.containsKey(key)) {
                result[i] = map.get(key);
            }
        }

        return result;
    }
}
